azure-keyvault/src/main/java/com/microsoft/azure/keyvault/messagesecurity/HttpMessageSecurity.java

```diff
@@ -0,0 +1,330 @@
+     * @param _serverSignatureKeyString
+     *      string with server signing key (public only) or null if not supported
+     * @param testMode
+     *      true for test mode (uses 0 for timestamp)
+     */
+    public HttpMessageSecurity(String _clientSecurityToken, String _clientEncryptionKeyString, String _clientSignatureKeyString, String _serverEncryptionKeyString, String _serverSignatureKeyString, boolean testMode) throws IOException{
+        this(_clientSecurityToken, _clientSignatureKeyString, _serverEncryptionKeyString, _serverSignatureKeyString);
+        this.testMode = testMode;
+        if (_clientEncryptionKeyString != null && !_clientEncryptionKeyString.equals("")){
+            this.clientEncryptionKey = MessageSecurityHelper.JsonWebKeyFromString(_clientEncryptionKeyString);
+        }
+    }
+
+    /**
+     * Protects existing request. Replaces its body with encrypted version.
+     *
+     * @param request
+     *      existing request.
+     *
+     * @return
+     *      new request with encrypted body if supported or existing request.
+     */
+    public Request protectRequest(Request request) throws IOException {
+        try{
+            Request result = request.newBuilder().header(AUTHENTICATE, BEARER_TOKEP_REFIX + clientSecurityToken).build();
+
+            if(!supportsProtection()) {
+                return result;
+            }
+
+            Buffer buffer = new Buffer();
+            request.body().writeTo(buffer);
+            String currentbody = buffer.readUtf8();
+
+            JsonWebKey clientPublicEncryptionKey = MessageSecurityHelper.GetJwkWithPublicKeyOnly(clientEncryptionKey);
+
+            String payload = currentbody.replace("}", ",\"rek\":{\"jwk\":" + clientPublicEncryptionKey.toString() + "}}");
+
+            JWEObject jweObject = protectPayload(payload);
+
+            JWSHeader jwsHeader = new JWSHeader("RS256",
+                    clientSignatureKey.kid(),
+                    clientSecurityToken,
+                    getCurrentTimestamp(),
+                    "PoP",
+                    null);
+
+            String jwsHeaderJsonb64 = MessageSecurityHelper.stringToBase64Url(jwsHeader.serialize());
+            String protectedPayload = MessageSecurityHelper.stringToBase64Url(jweObject.serialize());
+            byte[] data = (jwsHeaderJsonb64 + "." + protectedPayload).getBytes();
+
+            RsaKey clientSignatureRsaKey = new RsaKey(clientSignatureKey.kid(), clientSignatureKey.toRSA(true));
+            Pair<byte[], String> signature = clientSignatureRsaKey.signAsync(data, "RS256").get();
+
+            JWSObject jwsObject = new JWSObject(jwsHeader,
+                    protectedPayload,
+                    MessageSecurityHelper.bytesToBase64Url(signature.getKey()));
+
+            RequestBody body = RequestBody.create(MediaType.parse("application/jose+json"), jwsObject.serialize());
+
+            return result.newBuilder().post(body).build();
```

Comment:
>return result.newBuilder().post(body).build(); [](start = 12, length = 46)

You might consider getting the http method from the original request.  Right now the only APIs which support message security are post, but this could conceivably change.